File: autoviz/examples/python/21_tutorial_audio.py
#!/usr/bin/env python3
"""Publish foxglove.RawAudio (pcm-s16) for Autoviz Audio panel.

  /fake/audio  foxglove.RawAudio JSON

  /usr/bin/python3 examples/python/21_tutorial_audio.py
  ./build/autonomy/bin/autoviz
"""

from __future__ import annotations

from _reexec import early_reexec_if_needed

early_reexec_if_needed()

import argparse
import base64
import json
import math
import struct
import time

from _bootstrap import prepare_example_environment

prepare_example_environment()

import autolink

SR = 16000
CHUNK = 1600  # 100 ms @ 16 kHz
FREQ = 440.0


def make_pcm(phase0, n, freq=FREQ):
    """Return (pcm_bytes, next_phase)."""
    buf = bytearray()
    phase = phase0
    step = 2 * math.pi * freq / SR
    for _ in range(n):
        s = int(max(-32767, min(32767, 8000 * math.sin(phase))))
        buf += struct.pack('<h', s)
        phase += step
    return bytes(buf), phase


def make_msg(pcm, phase_unused=None):
    del phase_unused
    now = time.time_ns()
    return json.dumps(
        {
            'timestamp': {'sec': now // 1_000_000_000, 'nsec': now % 1_000_000_000},
            'format': 'pcm-s16',
            'sample_rate': SR,
            'number_of_channels': 1,
            'data': base64.b64encode(pcm).decode('ascii'),
        },
        separators=(',', ':'),
    ).encode()


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--rate', type=float, default=10.0)
    args = ap.parse_args()

    autolink.init('audio')
    n = autolink.Node('/autoviz/audio')
    w = n.create_writer('/fake/audio', 'foxglove.RawAudio', qos_depth=10)
    rate = autolink.Rate(args.rate)
    print(f'audio @ {args.rate} Hz → /fake/audio (pcm-s16 {SR} Hz)')

    phase = 0.0
    try:
        while not autolink.is_shutdown():
            pcm, phase = make_pcm(phase, CHUNK)
            w.write(make_msg(pcm))
            rate.sleep()
    except KeyboardInterrupt:
        pass
    finally:
        autolink.shutdown()


if __name__ == '__main__':
    main()

File: autoviz/examples/cpp/21_tutorial_audio.cpp
/******************************************************************************
 * Copyright 2026 The Openbot Authors (duyongquan)
 *****************************************************************************/

#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <iostream>
#include <memory>
#include <string>

#include "common/tutorial_utils.hpp"

namespace {

constexpr int kSr = 16000;
constexpr int kChunk = 1600;  // 100 ms @ 16 kHz
constexpr double kFreq = 440.0;

using autoviz::examples::MakeRawWriter;
using autoviz::examples::ParseRate;

std::string Base64Encode(const std::string& in) {
  static const char* kTbl =
      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  std::string out;
  out.reserve(((in.size() + 2) / 3) * 4);
  size_t i = 0;
  while (i + 2 < in.size()) {
    const uint32_t n = (static_cast<uint8_t>(in[i]) << 16) |
                       (static_cast<uint8_t>(in[i + 1]) << 8) |
                       static_cast<uint8_t>(in[i + 2]);
    out.push_back(kTbl[(n >> 18) & 63]);
    out.push_back(kTbl[(n >> 12) & 63]);
    out.push_back(kTbl[(n >> 6) & 63]);
    out.push_back(kTbl[n & 63]);
    i += 3;
  }
  if (i < in.size()) {
    uint32_t n = static_cast<uint8_t>(in[i]) << 16;
    out.push_back(kTbl[(n >> 18) & 63]);
    if (i + 1 < in.size()) {
      n |= static_cast<uint8_t>(in[i + 1]) << 8;
      out.push_back(kTbl[(n >> 12) & 63]);
      out.push_back(kTbl[(n >> 6) & 63]);
      out.push_back('=');
    } else {
      out.push_back(kTbl[(n >> 12) & 63]);
      out.push_back('=');
      out.push_back('=');
    }
  }
  return out;
}

std::string MakePcm(double* phase, int n, double freq = kFreq) {
  std::string buf;
  buf.resize(static_cast<size_t>(n) * 2);
  const double step = 2.0 * M_PI * freq / kSr;
  for (int i = 0; i < n; ++i) {
    int s = static_cast<int>(8000.0 * std::sin(*phase));
    if (s > 32767) s = 32767;
    if (s < -32767) s = -32767;
    const int16_t sample = static_cast<int16_t>(s);
    std::memcpy(&buf[static_cast<size_t>(i) * 2], &sample, 2);
    *phase += step;
  }
  return buf;
}

std::string MakeMsg(const std::string& pcm) {
  const auto now = std::chrono::system_clock::now().time_since_epoch();
  const auto ns =
      std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
  const int64_t sec = ns / 1000000000LL;
  const int64_t nsec = ns % 1000000000LL;
  return std::string("{\"timestamp\":{\"sec\":") + std::to_string(sec) +
         ",\"nsec\":" + std::to_string(nsec) +
         "},\"format\":\"pcm-s16\",\"sample_rate\":" + std::to_string(kSr) +
         ",\"number_of_channels\":1,\"data\":\"" + Base64Encode(pcm) + "\"}";
}

}  // namespace

int main(int argc, char** argv) {
  const double rate_hz = ParseRate(argc, argv, 10.0);

  if (!autolink::Init(argv[0])) return 1;
  auto node = autolink::CreateNode("/autoviz/audio");
  auto writer = MakeRawWriter(node, "/fake/audio", "foxglove.RawAudio");

  autolink::Rate rate(rate_hz);
  std::cout << "audio @ " << rate_hz << " Hz → /fake/audio (pcm-s16 " << kSr
            << " Hz)\n";

  double phase = 0.0;
  while (autolink::OK()) {
    const std::string pcm = MakePcm(&phase, kChunk);
    auto raw =
        std::make_shared<autolink::message::RawMessage>(MakeMsg(pcm));
    writer->Write(raw);
    rate.Sleep();
  }
  return 0;
}
